zuul.d/jobs.yaml

```diff
@@ -721,3 +721,18 @@
     cleanup-run: playbooks/new-tenant-juju/cleanup.yaml
     secrets:
       - serverstack_admin
+
+- job:
+    name: publish-charm
+    description: Publish a charm to charmhub
+    timeout: 3600
```

Comment:
An hour to publish a charm?

Reply:
> This job involves building the charm, so 20 minutes seems sane to me. What do you think?

That seems odd; we have a job called `charm-build` to do that.  Is it a duplication?

Reply:
The idea is to trigger this job once a patch is merged. So this job calls `charm-build` and `publish-charm` roles from the playbook `publish.yaml`.

Other way is to skip the charm-build and download *.charm from artifactory which is generated during Canonical CI run.

Reply:
It would be better, if possible, to use the same artifact that was created during the CI run; this would mean that what ends up in the charmhub is the same 'thing' as that which was tested, so the latter option (download) would be preferred I think.

Reply:
I agree that it would make sense to try and download the artifact created during the CI run, but I don't think that's the right answer here unless this build artifact came as a CI run that was built post-merge. The reason is that the CI may have run at a much different time than when the merge happens, and the $target_branch may now be different from what was built during the CI run.

Reply:
> but I don't think that's the right answer here unless this build artifact came as a CI run that was built post-merge.

This is probably true; it's possible that the charm artifact tested might miss something that's been merged since that successful test run.  I guess re-build is the only option here then.

Reply:
One more caveat to solve here is upload of resource oci-image and attaching during charm release.

* The existing publish-charm job should upload the charm, get the latest resource revisions for the corresponding track/risk and release the charm (attaching the resource revisions from earlier step). This job will be triggered as part of promote pipeline for every charm change-merge.
* Upload of resource oci-images should be separate job (triggered manually or periodically) that should perform functional tests before upload and attaching to the specific charm release.

Thoughts on this approach (Not sure how this is handled today for resources of file type)?

Reply:
Pushing the question back to you, @hemanthnakkina, which OCI image is the relevant branch testing against?  That's the OCI image that should be associated with the charm revision released to the channel.  Each track will have a 'known-good' OCI image for the payload that that track is testing against.  e.g. xena/{risk} will have a particular OCI image that is being tested against.  It sounds like this needs to either be in the charm itself (as part of the metadata of the charm -> pointing to an image) or held in config elsewhere (e.g. release-tools / the config section of it).  I think it probably should be in the charm's metadata as that means that the testing can carry that OCI image with it, and when it is published the image then gets associated for the deployment from charmhub.  Thoughts?

Reply:
Resources unfortunately do not have channels, they just have revisions. I suspect it was felt that revisions are 'good enough' for the resources, since you end up releasing a tuple into a channel (where the tuple is charm revision, resource X revision, resource Y revision, etc). This of course works fine as long as you have a single channel, but will eventually become _confusing_ when there are more channels involved. I suspect the long-term answer is "rocks" since they will also have channels. Today we can think of these as well defined tags that attached to an oci image in a registry somewhere (e.g. [the ubuntu/mysql oci image](https://hub.docker.com/r/ubuntu/mysql) on dockerhub).

I'm not sure how this works out in practice to be honest, but there needs to be some sort of indicator which revision of a resource should be used for a specific track. If we embed this within the charm, its much more clear - though it will mean more involved batches than what we have today; e.g.every single release we will have to go through all of the charms and update the revision of the resource which should be used - the revision will not only be unique per charm, but will need to be understood as to what the revision is.

This probably goes back to looking at the entire pipeline from image production to resource publication. On the one hand, there will be images that will need to be created out of the packages. Ideally it works such that when a package is placed into the archive pocket (straight into the archive on devel branches, into -proposed on stable branches) that a pipeline job is kicked off to build an oci image. The image validation will need to validate it using some sort of practical deployment and functional test, and the result of that (if successful) is the publishing of the oci image to *somewhere*. The key will be to tie the *somewhere* with the charm publishing/releasing that oci image as a unit for the revision that was tested.

Now this is a long digression to talk about the longer term plan for where this should go (imho), but I think it's an extremely important one in order to identify that there are multiple streams of work that need to converge/be fully considered for the releasing of a charm and corresponding resources into a channel. This may work well in a stable/released scenario but cause some headaches and challenges around releasing one that is in development (e.g. you need a charm change in order for the oci image resource validation to pass, but you need the new oci image resource in order for the charm change to pass - vicious cycle).

I think in the short term, the best path forward is to just publish a new resource on each merge - this will become more challenging with stable branches. On the charmhub side its essentially a new row in a database, but the same oci image is used if it doesn't actually change. If we can get the image hash from the charmhub api somehow, we can then be more intelligent about whether we need to upload the resource or not - but for now, let's just keep it simple.

Reply:
Bonus if you can only push if the image hash has actually changed from what is published.

The charmhub API does provide this in the output [docs here](https://api.snapcraft.io/docs/)

Reply:
I submitted a new patch which uploads oci-image for each charm change that means it creates a new resource revision for each change (same as above discussion, just reiterating).